### PR TITLE
Write interface for new serialization system

### PR DESCRIPTION
**Pull request by vtnerd**
This PR contains the write serialization interface, and some glue code from #8867 . This does **not** replace any existing serialization; the goal is to start a discussion on the design of the write system.

The write system - for epee binary - is roughly 56% faster than the existing design. This is primarily due to the removal of the intermediate DOM building stage - everything is written directly from objects to the "sink". I will work on JSON output numbers later.

The code for epee binary writing and json writing is not included, that can be reviewed in a future PR.

### CODE REVIEW

File: contrib/epee/include/serialization/wire/write.h
Comment (jeffro256) [inline contrib/epee/include/serialization/wire/write.h:74]: Is there any guarantee that `int` and `std::intmax_t` are different types on every system? Also, I haven't run benchmarks, but would it be a worthy tradeoff to crunch each `integer` and `unsigned_integer` virtual methods into one each? In other words, is the expanded vtables / extra binary size / worse caching worth not having to cast the integers?

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:74]: The only advantage is 32-bit systems. Theoretically some platform could use 128-bit integers on a 64-bit cpu (making the overload useful again), but that doesn't seem likely. It might be worth dropping to simplify things a bit more.

And I don't think its possible for `int` and `std::intmax_t` to have an issue, because both `long` and `long long` are required by the standard, and `std::intmax_t` must be the largest possible type on the platform. Even if they are all 32-bit, they should be unique types (afaik).

Of course, if I drop the overload entirely then it doesn't matter.

Comment (jeffro256) [reply contrib/epee/include/serialization/wire/write.h:74]: I think from a semantics point of view, since there is no difference between these two methods anyways that the overloads should be taken away IMO; it simplifies the interface. If there was a difference (see `serde`), then it would make sense to keep them in

File: contrib/epee/include/serialization/wire/field.h
Comment (jeffro256) [inline contrib/epee/include/serialization/wire/field.h:64]: Why would we ever want to enforce a required field when the array is empty in the current codebase?

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/field.h:64]: I'm not sure what you mean - when would want to make an empty array required? The only time would be if some external project required it for some reason. There's basically no code needed to support this feature, in fact supporting the empty-array-is-optional case is more code.

File: contrib/epee/include/serialization/wire/write.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h]: I removed the `int` and `unsigned` overloads, only the `intmax_t` and `uintmax_t` remain.

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h]: The unit tests just reminded me of one issue with removing the `unsigned` and `int` - when writing "through the base class" epee will _always_ write 64-bit integers. This can be prevented by checking the value and compressing the output (like my message pack implementation); the split previously allowed for a quick way to compress the types.

I think I'm going to make a change where the epee class always compresses the integer to the smallest possible (like message pack does now).

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h]: Nevermind, I remember why I cannot. MessagePack has a tag per array item, allowing you to compress individual integers, but epee does not. So if the base-class is used for writing epee data, then it will _always_ use 64-bit integers for outputs.

Comment (jeffro256) [reply contrib/epee/include/serialization/wire/write.h]: Also some binary formats (like bincode) have fixed size integer types. We might want to, in the future, add specific methods for this. Think like Rust's serde's `Serializer` interface with `serialize_u8`, `serialize_i16`, `serialize_u32`, `serialize_u64`, etc. By default, for most formats, these methods could feed directly into `integer()` and `unsigned_integer()`. But I guess we shouldn't over-engineer from the outset.

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h]: I'm not sure what you mean here - but MessagePack and Epee both have fixed integer types. Neither has variable size actually (whereas Protobuf does). That's not the issue. The problem is _every integer comes with its own tag_ for size with MessagePack, whereas with Epee there is one-fixed tag for every integer in an array.

I already wrote the Epee writer to serialize the exact fixed size when using the derived class (like serde), but it now fallsback to uint64 when using the base class. You can overload the write function multiple times if its a problem, so I don't see a need to engineer this further.

The MessagePack implement just compresses the value per integer, which is what most of the MessagePack serializers do.

Comment (jeffro256) [reply contrib/epee/include/serialization/wire/write.h]: What I mean is that in both of those cases, Epee and MessagePack, those formats are *self-describing*. You can determine the size of integers at runtime. Because of that property, you can be a little more willy-nilly casting integers to and forth into the format, and it isn't actually that important that we provide methods for every fixed size of integer. By contrast, in bincode, for example, this serialization information is assumed to be known beforehand, which saves space and computation, but if you don't know what size to deserialize an integer as, then reading from that point on might just give you garbage.

In MessagePack, since this tag is per-element (I'm assuming what you've told me I don't actually know anything about the Messagepack format), then you don't have to actually call a special `integer` method because the `writer` in this case could just assume you always want to compress if the value is small enough. And, likewise, since the `reader` can see the tags before deserializing the integers, we only need 2 methods for integers: signed and unsigned, and just cast everything accordingly.

I think it's fine like it is now with just `integer()/unsigned_integer()`, but if we wanted to support formats which use multiple sizes of integers but don't encode that integer size information in the data, then we would need more expressive interface with methods for each fixed integer size.

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h]: Yes, this is why I replaced everything but the internal crypto note serialization format.

File: contrib/epee/include/serialization/wire/write.h
Comment (jeffro256) [inline contrib/epee/include/serialization/wire/write.h:277]: Why is the `enable_if_t` needed here?

**Review by jeffro256 — COMMENTED** [review]
I'm getting very close to approval, just a couple of things left

File: contrib/epee/include/serialization/wire/write.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:152]: If a `write_bytes` is added to the `::wire::` namespace then this prevents ADL lookup. There are many such overloads, because any type that is not "owned" by our repo has to go into the `::wire::` namespace.

So you need a "clean" namespace (including global) of any `write_bytes` calls.

File: contrib/epee/include/serialization/wire/field.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/field.h:94]: I'm not sure what you mean here. You think specializations should be even easier than creating a wrapper type? I've done this in numerous places, and it didn't seem that bad in practice. Typically, I use a custom `struct` instead of this `boost::fusion::pair` object.

The other alternative is skipping usage of the `wire::object(...)` and instead call `start_object()` and `end_object()`, etc., directly. Its just that typically you'll need to handle out-of-order fields, and `wire::object(...)` does this for you.

File: contrib/epee/include/serialization/wire/write.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:128]: The problem is `bool` gets accepted for one of those types. There's probably a SFINAE method to fix that too, but I didn't bother coming up with it.

File: contrib/epee/include/serialization/wire/write.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:277]: There's an overload in `read.h` for the `wire::reader` case. I could omit this until the reader gets merged, but I think it's worth keeping in.

The overload allows for easy cpp usage - the `write_bytes` and `read_bytes` can forward to a single templated "map" function that calls `wire::object(...)` once for both the read and write case. i.e.

```c++
struct foo { int x; int y; }
void write_bytes(wire::writer&, const foo&);
void read_bytes(wire::reader&, foo&);

// .. cpp
template<typename F, typename T>
static void foo_map(F& format, T& self)
{ wire::object(format, WIRE_FIELD(x), WIRE_FIELD(y)); }

void write_bytes(wire::writer& dest, const foo& source) { foo_map(dest, source); }
void read_bytes(wire::reader& source, foo& dest) { foo_map(source, dest); } 
```

The boilerplate in the header and cpp can also further by macro'ed to make this a bit easier. This retains the "specify" once of the current method, while allowing for cpp files to reduce header bloat.

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:277]: And ! I forgot to mention why not `wire::object(reader& .. )` - templating allows forwarding of the derived type for performance reasons. Many of the calls within our codebase only actually use one type (p2p only uses epee, etc.), so specifying the derived makes more sense.

**Review by jeffro256 — APPROVED** [review]

File: contrib/epee/include/serialization/wire/write.h
Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:128]: I misspoke, I did this in the read interface, not sure why I didn't do it here. So this will be updated, along with the namespace recommendation and some squashing.

Comment (vtnerd) [reply contrib/epee/include/serialization/wire/write.h:128]: Nope, still getting the ambiguity errors, despite it being nearly identical to the read code. I'm not wasting further time on this, maybe in a future PR this can get cleaned up a bit.

**Comment by vtnerd** [comment]
I just force pushed a squashing commit, hopefully this won't be too hard to re-review.

**Review by selsta — APPROVED** [review]

**Pull request merged**
